[GTK] Use WebKitImage for the GTK snapshot implementation

spenap opened the pull request:
#### 4f90f08efe55ca33abe2635371f765229e18f64f
<pre>
[GTK] Use WebKitImage for the GTK snapshot implementation
<a href="https://bugs.webkit.org/show_bug.cgi?id=303173">https://bugs.webkit.org/show_bug.cgi?id=303173</a>

Reviewed by Adrian Perez de Castro.

This unifies the snapshot implementation to use WebKitImage when
2022_GLIB_API is enabled (converting from a WebKitImage to a GdkTexture
when using GTK, directly returning the WebKitImage for WPE), falling
back to cairo_surface_t otherwise.

It additional ensures we follow the same pixel format as WebImage, and
adds a test confirming that snapshot&apos;s colors are captured as intended.

Test: Tools/TestWebKitAPI/Tests/WebKitGLib/TestWebKitWebView.cpp
* Source/WebKit/UIProcess/API/glib/WebKitImage.cpp:
(webkit_image_class_init):
(webkitImageNew):
(webkitImageHash):
(webkitImageEqual):
(webkitImageLoad):
* Source/WebKit/UIProcess/API/glib/WebKitWebView.cpp:
* Tools/TestWebKitAPI/Tests/WebKitGLib/TestWebKitWebView.cpp:
(testWebViewColorQuadrants):
(beforeAll):

Canonical link: <a href="https://commits.webkit.org/304209@main">https://commits.webkit.org/304209@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/6b57deadfa5a29a507832dabfccf56ec770cdefc

| Misc | iOS, visionOS, tvOS & watchOS  | macOS  | Linux |  Windows |
| ----- | ---------------------- | ------- |  ----- |  --------- |
| [✅ 🧪 style](https://ews-build.webkit.org/#/builders/38/builds/134839 "Passed style check") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/159/builds/7273 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/138/builds/46089 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/5/builds/142355 "Built successfully") | [✅ 🛠 win](https://ews-build.webkit.org/#/builders/59/builds/86729 "Built successfully") 
| [✅ 🧪 bindings](https://ews-build.webkit.org/#/builders/9/builds/136709 "Passed tests") | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/155/builds/7888 "Built successfully") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/156/builds/7120 "Built successfully") | [✅ 🧪 wpe-wk2](https://ews-build.webkit.org/#/builders/34/builds/103036 "Passed tests") | [✅ 🧪 win-tests](https://ews-build.webkit.org/#/builders/60/builds/70327 "Passed tests") 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/11/builds/137786 "Passed tests") | [✅ 🧪 ios-wk2](https://ews-build.webkit.org/#/builders/162/builds/5567 "Passed tests") | [✅ 🧪 api-mac](https://ews-build.webkit.org/#/builders/18/builds/120864 "Passed tests") | [✅ 🧪 api-wpe](https://ews-build.webkit.org/#/builders/41/builds/83872 "Passed tests") | 
| | [✅ 🧪 ios-wk2-wpt](https://ews-build.webkit.org/#/builders/154/builds/5398 "Passed tests") | [✅ 🧪 api-mac-debug](https://ews-build.webkit.org/#/builders/165/builds/3015 "Passed tests") | [✅ 🛠 wpe-cairo-libwebrtc](https://ews-build.webkit.org/#/builders/166/builds/2946 "Built successfully") | 
| | [✅ 🧪 api-ios](https://ews-build.webkit.org/#/builders/13/builds/114613 "Passed tests") | [✅ 🧪 mac-wk1](https://ews-build.webkit.org/#/builders/135/builds/39015 "Passed tests") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/2/builds/145051 "Built successfully") | 
| | [✅ 🛠 vision](https://ews-build.webkit.org/#/builders/153/builds/6948 "Built successfully") | [✅ 🧪 mac-wk2](https://ews-build.webkit.org/#/builders/136/builds/39591 "Passed tests") | [✅ 🧪 gtk-wk2](https://ews-build.webkit.org/#/builders/1/builds/111416 "Passed tests") | 
| | [✅ 🛠 vision-sim](https://ews-build.webkit.org/#/builders/160/builds/7008 "Built successfully") | [✅ 🧪 mac-AS-debug-wk2](https://ews-build.webkit.org/#/builders/161/builds/5830 "Passed tests") | [✅ 🧪 api-gtk](https://ews-build.webkit.org/#/builders/21/builds/111753 "Passed tests") | 
| [✅ 🛠 🧪 merge](https://ews-build.webkit.org/#/builders/19/builds/28368 "Built successfully and passed tests") | [✅ 🧪 vision-wk2](https://ews-build.webkit.org/#/builders/164/builds/5236 "Passed tests") | [✅ 🧪 mac-wk2-stress](https://ews-build.webkit.org/#/builders/8/builds/117143 "Passed tests") | [✅ 🛠 playstation](https://ews-build.webkit.org/#/builders/134/builds/60879 "Built successfully") | 
| | [✅ 🛠 tv](https://ews-build.webkit.org/#/builders/158/builds/6995 "Built successfully") | [✅ 🧪 mac-intel-wk2](https://ews-build.webkit.org/#/builders/137/builds/35319 "Passed tests") | | 
| | [✅ 🛠 tv-sim](https://ews-build.webkit.org/#/builders/157/builds/6780 "Built successfully") | [✅ 🛠 mac-safer-cpp](https://ews-build.webkit.org/#/builders/120/builds/70574 "Built successfully") | | 
| | [✅ 🛠 watch](https://ews-build.webkit.org/#/builders/163/builds/7016 "Built successfully") | | | 
| | [✅ 🛠 watch-sim](https://ews-build.webkit.org/#/builders/152/builds/6890 "Built successfully") | | | 
<!--EWS-Status-Bubble-End-->

aperezdc commented on Source/WebKit/UIProcess/API/glib/WebKitWebView.cpp:
```diff
@@ -5186,33 +5179,32 @@ void webkit_web_view_get_snapshot(WebKitWebView* webView, WebKitSnapshotRegion r
+                    if (auto texture = adoptGRef(gdk_memory_texture_new(webkit_image_get_width(image.get()), webkit_image_get_height(image.get()), GDK_MEMORY_DEFAULT, bytes, webkit_image_get_stride(image.get())))) {
+                        g_task_return_pointer(task.get(), texture.leakRef(), g_object_unref);
+                        return;
+                    }
```
I think we should have an `else` part here, and use `g_task_return_new_error()` to notify client code that there was a failure, instead of leaving the `GTask` unresolved and potentially causing lockups if client code is waiting for the result.

spenap replied:
It's a bit messy with the guards, but when any of the implementations fail to return with `g_task_return_pointer`, they fall back to `g_task_return_new_error` in line 5210

aperezdc replied:
Indeed, you're right! Good that we are removing some preprocessor directives with this patch; but the diff having both the removed and changed ones made this difficult to notice.

aperezdc commented on Tools/TestWebKitAPI/Tests/WebKitGLib/TestWebKitWebView.cpp:
```diff
@@ -1169,6 +1170,83 @@ class SnapshotWebViewTest: public WebViewTest {
 #endif
 };
 
+#if PLATFORM(WPE)
```
It would be good to also run the test with `PLATFORM(GTK)`, but it wouldn't be too bad to leave that out or do it in a follow-up patch.

spenap replied:
I've done this before landing :-)

aperezdc reviewed (changes requested):
This is a neat simplification, and in general LGTM; but please take a look at the comment below about finalizing the `GTask` and update the patch accordingly. Thanks!

spenap committed:
[GTK] Use WebKitImage for the GTK snapshot implementation
https://bugs.webkit.org/show_bug.cgi?id=303173

Reviewed by Adrian Perez de Castro.

This unifies the snapshot implementation to use WebKitImage when
2022_GLIB_API is enabled (converting from a WebKitImage to a GdkTexture
when using GTK, directly returning the WebKitImage for WPE), falling
back to cairo_surface_t otherwise.

It additional ensures we follow the same pixel format as WebImage, and
adds a test confirming that snapshot's colors are captured as intended.

Test: Tools/TestWebKitAPI/Tests/WebKitGLib/TestWebKitWebView.cpp
* Source/WebKit/UIProcess/API/glib/WebKitImage.cpp:
(webkit_image_class_init):
(webkitImageNew):
(webkitImageHash):
(webkitImageEqual):
(webkitImageLoad):
* Source/WebKit/UIProcess/API/glib/WebKitWebView.cpp:
* Tools/TestWebKitAPI/Tests/WebKitGLib/TestWebKitWebView.cpp:
(testWebViewColorQuadrants):
(beforeAll):

Canonical link: https://commits.webkit.org/304209@main
